### DISCUSSION TITLE
What's the proper way to do async work in OnFrameworkInitializationCompleted?

### DISCUSSION THREAD

**Discussion by rfvgyhn** [AvaloniaUI/Avalonia#13161] [Q&A] [answered]
After creating an app with `dotnet new avalonia.mvvm` (v11.0.2) and awaiting a task in `OnFrameworkInitializationCompleted()`, no window appears/is created.

```diff
- public override void OnFrameworkInitializationCompleted()
+ public override async void OnFrameworkInitializationCompleted()
{
    if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
    {
+       await Task.Delay(1);
        desktop.MainWindow = new MainWindow
        {
            DataContext = new MainWindowViewModel(),
        };
    }

    base.OnFrameworkInitializationCompleted();
}
```

If I manually call `.Show`, the window does appear. Is that the correct way to go about it?

```diff
public override async void OnFrameworkInitializationCompleted()
{
    if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
    {
        await Task.Delay(1);
        desktop.MainWindow = new MainWindow
        {
            DataContext = new MainWindowViewModel(),
        };
        
+       desktop.MainWindow.Show();
    }

    base.OnFrameworkInitializationCompleted();
}
```

Or is blocking or moving the async work somewhere else the correct way?

```sh
$ dotnet --info
.NET SDK:
 Version:   7.0.401
 Commit:    eb26aacfec

Runtime Environment:
 OS Name:     arch
 OS Version:  
 OS Platform: Linux
 RID:         arch-x64
 Base Path:   /usr/share/dotnet/sdk/7.0.401/

Host:
  Version:      8.0.0-rc.1.23419.4
  Architecture: x64
  Commit:       92959931a3
  RID:          linux-x64

.NET SDKs installed:
  7.0.401 [/usr/share/dotnet/sdk]
  8.0.100-rc.1.23463.5 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 7.0.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 7.0.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.1.23419.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/usr/share/dotnet]

global.json file:
  /home/user/dev/Avalonia11AsyncVoidTest/global.json
```

global.json
```json
{
  "sdk": {
    "version": "7.0.401"
  }
}
```

**Comment by 13xforever** [comment, marked as answer]
you're blocking the ui thread between initialization and presentation (form show) actions, so this should be the expected outcome

you either make absolutely sure you do no long ops here, or you schedule an async op on another thread and return immediately, but at this point you might as well just use form show event for your main view

**Comment by thevortexcloud** [comment]
Personally I think you should not be doing async code here. Although I guess it really depends on what you are doing. If it's logic that can be moved into a view/viewmodel you should.

**Comment by timunie** [comment]
You can also call an async Task with FireAndForget. That way the view shows up while data is being loaded. In the end of your task, populate the ViewModel with your data. 

For example: https://github.com/brminnick/AsyncAwaitBestPractices

**Reply by crazyfox55** [reply]
I don't really understand how the fire and forget method will set the data on the view model? Does it need to be dispatched on the UI thread some how otherwise the property changed method will cause the UI to update on the wrong thread?

**Reply by timunie** [reply]
`Dispatcher.UiThread.Post()` can be used
